src/hex_repo.erl

```diff
@@ -0,0 +1,204 @@
+-module(hex_repo).
+-export([
+    default_options/0,
+    get_names/0,
+    get_names/1,
+    get_versions/0,
+    get_versions/1,
+    get_package/1,
+    get_package/2,
+    get_tarball/2,
+    get_tarball/3
+]).
+
+-type options() :: [{client, client()} | {repo, repo()} | {verify, boolean()}].
+-type client() :: #{adapter => hex_http:adapter(), user_agent_string => string()}.
+-type repo() :: #{uri => string(), public_key => binary()}.
+
+%%====================================================================
+%% API functions
+%%====================================================================
+
+%% @doc
+%% Default options used to interact with the repository.
+%% @end
+-spec default_options() -> options().
+default_options() ->
+    %% https://hex.pm/docs/public_keys
+    PublicKey = <<"-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEApqREcFDt5vV21JVe2QNB
+Edvzk6w36aNFhVGWN5toNJRjRJ6m4hIuG4KaXtDWVLjnvct6MYMfqhC79HAGwyF+
+IqR6Q6a5bbFSsImgBJwz1oadoVKD6ZNetAuCIK84cjMrEFRkELtEIPNHblCzUkkM
+3rS9+DPlnfG8hBvGi6tvQIuZmXGCxF/73hU0/MyGhbmEjIKRtG6b0sJYKelRLTPW
+XgK7s5pESgiwf2YC/2MGDXjAJfpfCd0RpLdvd4eRiXtVlE9qO9bND94E7PgQ/xqZ
+J1i2xWFndWa6nfFnRxZmCStCOZWYYPlaxr+FZceFbpMwzTNs4g3d4tLNUcbKAIH4
+0wIDAQAB
+-----END PUBLIC KEY-----">>,
+    Client = #{adapter => hex_http_httpc, user_agent_string => "(httpc)"},
+    Repo = #{uri => "https://repo.hex.pm", public_key => PublicKey},
+    [{client, Client}, {repo, Repo}, {verify, true}].
+
+%% @doc
+%% Gets names resource from the repository.
+%%
+%% Same as `hex_repo:get_names(hex_repo:default_options())'.
+%%
+%% Examples:
+%%
+%% ```
+%%     hex_repo:get_names().
+%%     %%=> {ok, #{package => [
+%%     %%=>     #{name => <<"package1">>},
+%%     %%=>     #{name => <<"package2">>},
+%%     %%=>     ...]}}
+%% '''
+%% @end
+-spec get_names() -> {ok, map()} | {error, term()}.
+get_names() ->
+    get_names(default_options()).
+
+%% @doc
+%% Gets names resource from the repository.
+%%
+%% See `get_names/1' for examples.
+%% @end
+-spec get_names(options()) -> {ok, map()} | {error, term()}.
```

Comment:
this API proved to be a little bit awkward to use, if we want to overwrite an option we need to remember to pass default_options (otherwise we get bad errors like repo key is undefined etc):

```erlang
hex_repo:get_names([{adapter, xhttp} | hex_repo:default_options()).
```

instead of:

```erlang
hex_repo:get_names([{adapter, xhttp}])
```

@ericmj I'm gonna go with the latter and mention in the docs that we are overwriting default_options but let me know if you have other ideas.

Reply:
Sounds good 👍